src/kernels/hip_f8_impl.hpp

```diff
@@ -0,0 +1,316 @@
+    }
+    mantissa >>= (mfmt - wm);
+    mantissa &= (1 << wm) - 1;
+    if(exponent == 31)
+        return (sign << 7) | 0x7B;
+    return (sign << 7) | (exponent << wm) | mantissa;
+}
+
+template <int wm, int we, typename T, bool negative_zero_nan, bool clip>
+MIOPEN_HIP_HOST_DEVICE uint8_t cast_to_f8(T _x, bool stoch, uint32_t rng)
+{
+    constexpr bool is_half  = std::is_same<T, half>::value;
+    constexpr bool is_float = std::is_same<T, float>::value;
+    static_assert(wm + we == 7, "wm+we==7");
+    static_assert(is_half || is_float, "Only half and float can be cast to f8");
+
+    if(sizeof(T) == 2 && we == 5 && !negative_zero_nan)
+        return cast_to_f8_no_range_reduce<2, 5, half>(static_cast<half>(_x), stoch, rng);
+
+    const int mfmt = (sizeof(T) == 4) ? 23 : 10;
+    uint32_t x;
+    if(sizeof(T) == 4)
+        x = *(reinterpret_cast<uint32_t*>(&_x)); // cppcheck-suppress invalidPointerCast
+    else
+        x = *(reinterpret_cast<uint16_t*>(&_x)); // cppcheck-suppress invalidPointerCast
+
+    uint32_t head, mantissa;
+    int exponent;
+    uint32_t sign;
+
+    if(sizeof(T) == 4)
+    {
+        head     = x & 0xFF800000;
+        mantissa = x & 0x7FFFFF;
+        exponent = (head >> 23) & 0xFF;
+        sign     = head >> 31;
+    }
+    else
+    {
+        head     = x & 0xFC00;
+        mantissa = x & 0x3FF;
+        exponent = (head >> 10) & 0x1F;
+        sign     = head >> 15;
+    }
+
+    uint32_t signed_inf = (sign << 7) + (((1 << we) - 1) << wm);
+
+    if(negative_zero_nan)
+    {
+        if(sizeof(T) == 4)
+        {
+            if((x & 0x7F800000) == 0x7F800000)
+                return 0x80;
+        }
+        else
+        {
+            // if(__hisinf(x) || __hisnan(x))
+            if((x & 0x7C00) == 0x7C00)
+                return 0x80;
+        }
+    }
+    else
+    {
+        if(sizeof(T) == 4)
+        {
+            if((x & 0x7F800000) == 0x7F800000)
+                return signed_inf + (mantissa != 0 ? 1 : 0);
+        }
+        else
+        {
+            if((x & 0x7C00) == 0x7C00)
+                return signed_inf + (mantissa != 0 ? 1 : 0);
+        }
+    }
+    if(x == 0)
+        return 0;
+
+    if(is_half && we == 5 && negative_zero_nan && exponent == 0)
+    {
+        exponent += 1;
+        // TODO: call __clz when this is device code
+        int sh = 1 + __builtin_clz(mantissa) - (32 - mfmt);
+        mantissa <<= sh;
+        exponent -= sh;
+        /*
+        while(mantissa < (1<<mfmt)) {
+          mantissa <<= 1;
+          exponent -= 1;
+        }
+        */
+        mantissa &= ~(1 << mfmt);
+    }
+
+    uint32_t drop_mask = (1 << (mfmt - wm)) - 1;
+    const int max_exp  = (1 << we) - (negative_zero_nan ? 1 : 2);
+    const int exp_low_cutoff =
+        (sizeof(T) == 4 ? 128 : 16) - (1 << (we - 1)) + 1 - (negative_zero_nan ? 1 : 0);
+
+    exponent -= exp_low_cutoff - 1;
+    if(exponent <= 0)
+        drop_mask = (1 << (mfmt - wm + 1 - exponent)) - 1;
```

Comment:
@JehandadKhan some other tests has shown the shift may be too large in some cases: e.g.
```
runtime error: shift exponent 141 is too large for 32-bit type 'int'
```